File: Partner API V2/Ruby/requestonescansession.rb
require "webrick"
require "json"
require "net/http"
require "securerandom"

require_relative "onescanhmac"
require_relative "onescanconfig"

module Onescan

  class RequestOnescanSession < WEBrick::HTTPServlet::AbstractServlet

    def do_POST request, response

      onescanRequest = {
        :ProcessType => "Payment",
        :MessageType => "StartPayment",
        # TODO: For WebContent Processes change the process and messagetypes:
        #:ProcessType => "WebContent",
        #:MessageType => "WebContent",
        # TODO: For Login Processes:
        #:ProcessType => "Login",
        #:MessageType => "StartLogin",
        #:LoginPayload => {
        #  :FriendlyName => "Demo site",
        #  :SiteIdentifier => "YOUR UNIQUE SITE ID",
        #  # TODO: Other LoginModes are: TokenOrCredentials, UserToken, TokenOrCredentials, Register
        #  :LoginMode => "UsernamePassword",
        #  # TODO: Profiles are required for Register (and with UserToken when the response MessageType "RegisterUser" is used.
        #  :Profiles => [ "basic" ]
        #},
        # The Session Data property can be set by you to represent a hold data relating
        # to the users session in some way and will be passed back with each callback
        # so you can locate the session (eg Order number) that the request belongs to.
        :SessionData => "CUSTOM SESSION DATA",
        :Version => 2,
        :MetaData => {
          :EndpointURL => Onescan::CALLBACK_URL
        }
      }

      # TODO: You can set the PurchasePayload early if you wish (see StartPayment callback)

      onescanRequestMessage = JSON.generate(onescanRequest)

      uri = URI(Onescan::SERVER_URL)
      req = Net::HTTP::Post.new(uri)

      req.body = onescanRequestMessage
      req.content_type = "application/json"
      req["x-onescan-account"] = Onescan::ACCOUNT_KEY
      # Sign the header
      hmac = Onescan::HMAC.encode(onescanRequestMessage, Onescan::SECRET)
      req["x-onescan-signature"] = hmac

      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      # ca_file may be required if certs are not present on server
      http.ca_file = "./curl/curl-ca-bundle.crt"

      res =  http.request(req)
      responseData = res.body

      # validate the response header signature
      onescanSignature = res["x-onescan-signature"]
      Onescan::HMAC.validateSignature(responseData,Onescan::SECRET,onescanSignature)

      response.status = 200
      response["Content-Type"] = "application/json"
      response.body = responseData
    end
  end

end

File: Partner API V2/PHP/RequestOnescanSession.php
<?php
  include("OnescanConfig.php");
  include("HMAC.php");

  $onescanRequest = NULL;
  // Build the Request Onescan Session payload
  $onescanRequest->ProcessType = "Payment";
  $onescanRequest->MessageType = "StartPayment";
  // TODO: For WebContent Processes change the process and messagetypes:
  //$onescanRequest->ProcessType = "WebContent";
  //$onescanRequest->MessageType = "WebContent";
  // TODO: For Login Processes:
  //$onescanRequest->ProcessType = "Login";
  //$onescanRequest->MessageType = "StartLogin";
  //$onescanRequest->LoginPayload = createLoginRequest();

  // The Session Data property can be set by you to represent a hold data relating
  // to the users session in some way and will be passed back with each callback
  // so you can locate the session (eg Order number) that the request belongs to.
  $onescanRequest->SessionData = "YOUR_SESSION_DATA";
  $onescanRequest->Version = 2;
  $onescanRequest->MetaData->EndpointURL = $config->OnescanCallbackURL;

  // TODO: For payment processes you can set the PurchasePayload early if you wish (see StartPayment callback)

  // JSon Encode
  $onescanRequestMessage = json_encode($onescanRequest);

  // Sign the header
  $hmac = hashit($onescanRequestMessage,$config->OnescanSecret);
  $header = array(
      "content-type: application/json",
      "x-onescan-account: " . $config->OnescanAccountKey,
      "x-onescan-signature: " . $hmac
  );
  $uri = curl_init($config->OnescanServerURL);
  curl_setopt($uri, CURLOPT_HTTPHEADER, $header);
  curl_setopt($uri,CURLOPT_POST,true);
  curl_setopt($uri,CURLOPT_RETURNTRANSFER,true);
  curl_setopt($uri, CURLOPT_POSTFIELDS, $onescanRequestMessage);
  // TODO: For windows you might need to provide the curl cert bundle
  curl_setopt($uri, CURLOPT_CAINFO, dirname(__FILE__) . "/curl/curl-ca-bundle.crt");

  try {
    $response = curl_exec($uri);
    if (curl_errno($uri)) {
        throw new Exception("HTTP Error " . curl_errno($uri) . ". " . curl_error($uri));
    }
    curl_close($uri);
  }
  catch (Exception $e) {
      curl_close($uri);
      file_put_contents("php://stdout", $hmac . "ffff\n");
      throw $e;
  }

  // Check that the signature of the header matches the payload
  validateSignature($response,$config->OnescanSecret);

  echo $response;


  function createLoginRequest()
  {
      $loginPayload = NULL;
      $loginPayload->FriendlyName = "Demo site";
      $loginPayload->SiteIdentifier = "YOUR UNIQUE SITE ID";
      //TODO: Other LoginModes are: TokenOrCredentials, UserToken, TokenOrCredentials, Register
      $loginPayload->LoginMode = "UsernamePassword";

      //TODO: Profiles are required for Register (and with UserToken when the response MessageType "RegisterUser" is used.
      $loginPayload->Profiles = [ "basic" ];
      return $loginPayload;
  }
?>
